feat!(*): use required name attribute instead of block label

hainenber opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description

#### Which issue(s) this PR fixes

<!-- Uncomment the following line if you want that GitHub issue gets automatically closed after merging the PR -->
Fixes #5990 

#### Notes to the Reviewer

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [x] Config converters updated

clayton-cornell commented on docs/sources/flow/reference/components/prometheus.exporter.cloudwatch.md:
```diff
@@ -249,6 +251,7 @@ You can configure the `static` block one or multiple times to scrape metrics wit
 | Name          | Type           | Description                                                                                                                                                  | Default | Required |
 | ------------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------- | -------- |
 | `regions`     | `list(string)` | List of AWS regions.                                                                                                                                         |         | yes      |
+| `name`        | `string` 		 | Name to define current batch of scrapped CloudWatch metrics.                                                                                                                                         |         | yes      |
```
```suggestion
| `name`        | `string` 		 | Name to define the current batch of scrapped CloudWatch metrics.                                                                                                                                         |         | yes      |
```

clayton-cornell commented on docs/sources/flow/reference/components/prometheus.exporter.snmp.md:
```diff
@@ -66,11 +67,12 @@ The following blocks are supported inside the definition of
 
 The `target` block defines an individual SNMP target.
-The `target` block may be specified multiple times to define multiple targets. The label of the block is required and will be used in the target's `job` label.
+The `target` block may be specified multiple times to define multiple targets. `name` attribute is required and will be used in the target's `job` label.
```
```suggestion
You can specify the `target` block multiple times to define multiple targets. The `name` attribute is required and is used in the target's `job` label.
```

clayton-cornell commented on docs/sources/flow/reference/components/pyroscope.scrape.md:
```diff
@@ -366,25 +366,27 @@ The following arguments are supported:
 
 The `profile.custom` block allows for collecting profiles from custom
-endpoints. Blocks must be specified with a label:
+endpoints. Blocks must be specified with `name` attribute:
```
```suggestion
endpoints. Blocks must be specified with a `name` attribute:
```

clayton-cornell commented on docs/sources/flow/reference/components/pyroscope.scrape.md:
````diff
@@ -366,25 +366,27 @@ The following arguments are supported:
 ```
 
-Multiple `profile.custom` blocks can be specified. Labels assigned to
+Multiple `profile.custom` blocks can be specified. `name` attribute assigned to
````
```suggestion
Multiple `profile.custom` blocks can be specified. The `name` attribute assigned to
```

clayton-cornell commented on docs/sources/flow/release-notes.md:
```diff
@@ -41,6 +41,149 @@ This functionality is now only available in the main configuration.
 
+### Breaking change: Label for children blocks of `prometheus.exporter.cloudwatch`, `prometheus.exporter.snmp` and `pyroscope.scrape` are removed.
+
+Previously in aforementioned components, there are children blocks that requires labels which would be used for naming. 
```
```suggestion
Previously, the child blocks in the `prometheus.exporter.cloudwatch`, `prometheus.exporter.snmp`, and `pyroscope.scrape components required labels which would be used for naming. 
```

Was this all child blocks? Or just the ones in the table below?

hainenber replied:
Only the ones in the table

clayton-cornell commented on docs/sources/flow/release-notes.md:
```diff
@@ -41,6 +41,149 @@ This functionality is now only available in the main configuration.
+### Breaking change: Label for children blocks of `prometheus.exporter.cloudwatch`, `prometheus.exporter.snmp` and `pyroscope.scrape` are removed.
+
+Previously in aforementioned components, there are children blocks that requires labels which would be used for naming. 
+More information in the following table.
+
```
```suggestion

```

We may need to restructure this a little depending on the answer to the previous question.

clayton-cornell commented on docs/sources/flow/release-notes.md:
```diff
@@ -41,6 +41,149 @@ This functionality is now only available in the main configuration.
+Previously in aforementioned components, there are children blocks that requires labels which would be used for naming. 
+More information in the following table.
+
+| Component                        | Affected blocks        |
```
```suggestion
| Component                        | Child blocks        |
```

clayton-cornell commented on docs/sources/flow/release-notes.md:
```diff
@@ -41,6 +41,149 @@ This functionality is now only available in the main configuration.
+| `prometheus.exporter.snmp`       | `target`, `walk_param` |
+| `pyroscrope.scrape`              | `profile.custom`       |
+
+In the new version, user needs to be specify `name` attribute instead, which allow less restrictive naming.
```
```suggestion
Now, you must specify a `name` attribute. This new attribute allows less restrictive naming.
```

clayton-cornell commented on docs/sources/flow/release-notes.md:
````diff
@@ -41,6 +41,149 @@ This functionality is now only available in the main configuration.
+
+In the new version, user needs to be specify `name` attribute instead, which allow less restrictive naming.
+
+Old configuration example:
+```river
````
```suggestion
Old configuration example:

```river
```

clayton-cornell commented on docs/sources/flow/release-notes.md:
````diff
@@ -41,6 +41,149 @@ This functionality is now only available in the main configuration.
+}
+```
+
+New configuration example:
+```river
````
```suggestion
New configuration example:

```river
```

clayton-cornell reviewed (commented):
Some doc suggestions

clayton-cornell commented on CHANGELOG.md:
```diff
@@ -16,6 +16,12 @@ Main (unreleased)
 
 - For `otelcol.exporter` components, change the default value of `disable_high_cardinality_metrics` to `true`. (@ptodev)
 
+- The children blocks in following components requires a mandatory `name`
+  argument instead of a block label. (@hainenber)
```
```suggestion
- The child blocks in following components require a mandatory `name` argument instead of a block label. (@hainenber)
```

clayton-cornell commented on docs/sources/flow/release-notes.md:
```diff
@@ -41,11 +41,157 @@ This functionality is now only available in the main configuration.
+### Breaking change: Label for children blocks of `prometheus.exporter.cloudwatch`, `prometheus.exporter.snmp` and `pyroscope.scrape` are removed.
+
+Previously in aforementioned components, there are children blocks that requires labels which would be used for naming. 
+More information in the following table.
```
```suggestion
Now, you must specify a `name` attribute for the following child blocks.
```

What about moving the text from line 55 up to before the table and describing it this way? I dropped the "less restrictive" part. It's not really adding a lot... unless it's a really important point.

clayton-cornell commented on docs/sources/flow/release-notes.md:
```diff
@@ -41,11 +41,157 @@ This functionality is now only available in the main configuration.
+| `prometheus.exporter.snmp`       | `target`, `walk_param` |
+| `pyroscrope.scrape`              | `profile.custom`       |
+
+Now, you must specify a `name` attribute. This new attribute allows less restrictive naming.
+
```
```suggestion

```

If we move this text above the table, we can delete this line.

clayton-cornell reviewed (commented):
Some additional doc suggestions

hainenber commented:
I'll close this PR off as Grafana folks are understandably refrained from making breaking changes since v1.0 is pretty close. And this is multiple breaking changes in 1 PR!
